File: leetcode/DynamicProgram/leetcode-0063/63.go
package arrays

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	if obstacleGrid == nil || len(obstacleGrid) == 0 || len(obstacleGrid[0]) == 0 {
		return 0
	}

	if obstacleGrid[0][0] > 0 {
		return 0
	}

	row, col := len(obstacleGrid), len(obstacleGrid[0])
	dp := make([]int, col+1)
	dp[1] = 1
	for i := 1; i < col; i++ {
		if obstacleGrid[0][i] > 0 {
			dp[i+1] = 0
		} else {
			dp[i+1] = dp[i]
		}
	}

	for i := 1; i < row; i++ {
		for j := 0; j < col; j++ {
			if obstacleGrid[i][j] > 0 {
				dp[j+1] = 0
			} else {
				dp[j+1] += dp[j]
			}
		}
	}
	return dp[col]
}

File: leetcode/DynamicProgram/leetcode-0063/63.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid.length == 0 || obstacleGrid[0].length == 0) {
            return 0;
        }
        if (obstacleGrid[0][0] > 0) {
            return 0;
        }
        int row = obstacleGrid.length;
        int col = obstacleGrid[0].length;
        int[] dp = new int[col + 1];
        dp[1] = 1;
        for (int i = 2; i <= col; i++) {
            if (obstacleGrid[0][i - 1] > 0) {
                dp[i] = 0;
            } else {
                dp[i] = dp[i - 1];
            }
        }

        for (int i = 1; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (obstacleGrid[i][j] > 0) {
                    dp[j + 1] = 0;
                } else {
                    dp[j + 1] += dp[j];
                }
            }
        }
        return dp[col];
    }
}